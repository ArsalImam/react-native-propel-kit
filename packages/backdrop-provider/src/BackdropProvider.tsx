import React, {useCallback, useRef, FunctionComponent, useState, ReactNode, useEffect, useMemo} from 'react';
import {Animated, Easing, StyleSheet, EasingFunction, Platform} from 'react-native';

export type Props = {
  backgroundColor?: string;
  children?: ReactNode;
  duration?: number;
  easing?: EasingFunction;
  opacity?: number;
  useNativeDriver?: boolean;
  zIndex?: number;
};

const IOS_OPACITY = 0.4; // @NOTE from native ActionSheet backdrop

export const defaultProps = {
  backgroundColor: 'black',
  duration: 300,
  easing: Easing.inOut(Easing.ease),
  opacity: IOS_OPACITY,
  useNativeDriver: true,
  zIndex: 99
};

const BackdropProvider: FunctionComponent<Props> = ({
  backgroundColor = 'black',
  children,
  duration = defaultProps.duration,
  easing = defaultProps.easing,
  opacity = defaultProps.opacity,
  useNativeDriver = defaultProps.useNativeDriver,
  zIndex = defaultProps.zIndex
}) => {
  const [isVisible, setIsVisible] = useState<boolean>(false);
  const [displayStyle, setDisplayStyle] = useState<'flex' | 'none'>('flex');
  const latestIsVisible = useRef<boolean>(isVisible);

  // Show backdrop
  const show = useCallback(() => {
    setIsVisible(true);
    setDisplayStyle('flex');
  }, []);

  // Hide backdrop
  const hide = useCallback(() => {
    setIsVisible(false);
    latestIsVisible.current = false;
  }, []);

  // Track isVisible latest value with a ref
  useEffect(() => {
    latestIsVisible.current = isVisible;
  }, [isVisible]);

  // Toggle backdrop
  // @NOTE a ref is used to provide a stable context accross renders
  const toggle = useCallback(() => {
    !latestIsVisible.current ? show() : hide();
  }, [latestIsVisible, show, hide]);

  // Expose API via context
  const contextValue = useMemo(() => {
    return {show, hide, toggle};
  }, [show, hide, toggle]);

  // Animate backdrop opacity along "isVisible" prop
  const animatedOpacity = useRef(new Animated.Value(0));
  useEffect(() => {
    const animation = Animated.timing(animatedOpacity.current, {
      duration: isVisible ? duration : 200,
      easing,
      toValue: isVisible ? opacity : 0,
      useNativeDriver
    });
    animation.start(() => {
      setDisplayStyle(isVisible ? 'flex' : 'none');
    });
    return () => animation.stop();
  }, [isVisible, duration, easing, opacity, useNativeDriver]);

  return (
    <BackdropContext.Provider value={contextValue}>
      {children}
      <Animated.View
        style={[
          StyleSheet.absoluteFill,
          // @NOTE hack around bug https://github.com/facebook/react-native/issues/18415
          Platform.select({
            android: {
              position: displayStyle === 'none' ? 'relative' : 'absolute'
            }
          }),
          {display: displayStyle, zIndex, backgroundColor}
        ]}
        opacity={animatedOpacity.current}
      />
    </BackdropContext.Provider>
  );
};

export default BackdropProvider;

export type ContextProps = null | {
  show: () => void;
  hide: () => void;
  toggle: () => void;
};

export const BackdropContext = React.createContext<ContextProps>(null);
