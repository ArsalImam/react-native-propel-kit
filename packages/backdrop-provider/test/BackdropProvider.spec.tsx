// import React from 'react';
// import {StyleSheet} from 'react-native';
// import {shallow} from 'enzyme';

import React from 'react';
import 'react-native';
import {create} from 'react-test-renderer';

import BackdropProvider from './../src';

describe('<BackdropProvider>', () => {
  it('renders correctly', () => {
    const component = create(<BackdropProvider />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});

// const component = create(<ContextProvider />);
// const instance = component.getInstance();

// expect(instance.getDose()).toBe(5);
// expect(instance.state.productData.dose).toBe("5 mg");

// instance.setDose(2);

// expect(instance.getDose()).toBe(2);
// expect(instance.state.productData.dose).toBe("2 mg");

// const noop = () => {};
// const sampleText = 'foo bar baz';

// describe('<BackdropProvider>', () => {
//   it('should match snapshot', () => {
//     const wrapper = shallow(<BackdropProvider />);
//     expect(wrapper).toMatchSnapshot();
//   });

// TextInput props
// describe('onTouch prop', () => {
//   it('should properly call callback', () => {
//     const onTouchSpy = jest.fn();
//     const component = shallow(<DatePicker onTouch={onTouchSpy} />);
//     component.find('DatePickerInput').simulate('touch');
//     expect(onTouchSpy).toHaveBeenCalledTimes(1);
//   });
// });
// describe('onFocus prop', () => {
//   it('should properly call callback', () => {
//     const onFocusSpy = jest.fn();
//     const component = shallow(<DatePicker onFocus={onFocusSpy} />);
//     component.find('DatePickerInput').simulate('focus');
//     expect(onFocusSpy).toHaveBeenCalledTimes(1);
//   });
// });

//   it('renders external container style from object', () => {
//     const Clone = React.cloneElement(Component, {style: {backgroundColor: 'red'}});
//     const wrapper = shallow(Clone);
//     expect(wrapper).toMatchSnapshot();
//   });

//   it('renders external container style from ID', () => {
//     const styles = StyleSheet.create({
//       containerStyle: {
//         padding: 4
//       }
//     });
//     const Clone = React.cloneElement(Component, {style: styles.containerStyle});
//     const wrapper = shallow(Clone);
//     expect(wrapper).toMatchSnapshot();
//   });

//   it('renders external container style from array', () => {
//     const style1 = {backgroundColor: 'yellow'};
//     const styles = StyleSheet.create({
//       style2: {
//         maxHeight: 32
//       }
//     });
//     const Clone = React.cloneElement(Component, {style: [style1, styles.style2]});
//     const wrapper = shallow(Clone);
//     expect(wrapper).toMatchSnapshot();
//   });

//   it('renders external text style from object', () => {
//     const Clone = React.cloneElement(Component, {textStyle: {color: 'black'}});
//     const wrapper = shallow(Clone);
//     expect(wrapper).toMatchSnapshot();
//   });

//   it('renders external text style from ID', () => {
//     const styles = StyleSheet.create({
//       textStyle: {
//         fontSize: 12
//       }
//     });
//     const Clone = React.cloneElement(Component, {textStyle: styles.textStyle});
//     const wrapper = shallow(Clone);
//     expect(wrapper).toMatchSnapshot();
//   });

//   it('renders external text style from array', () => {
//     const style1 = {color: 'blue'};
//     const styles = StyleSheet.create({
//       style2: {
//         fontFamily: 'sans-serif'
//       }
//     });
//     const Clone = React.cloneElement(Component, {textStyle: [style1, styles.style2]});
//     const wrapper = shallow(Clone);
//     expect(wrapper).toMatchSnapshot();
//   });
// });

// describe('Functionality', () => {
//   it('handles onPress callback', () => {
//     const onPressSpy = jest.fn();
//     const Clone = React.cloneElement(Component, {onPress: onPressSpy});
//     const wrapper = shallow(Clone);
//     wrapper.simulate('press');
//     expect(onPressSpy).toHaveBeenCalledTimes(1);
//   });
// });
