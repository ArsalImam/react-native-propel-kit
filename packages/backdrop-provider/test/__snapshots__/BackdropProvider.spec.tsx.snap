// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BackdropProvider> renders correctly 1`] = `
<View
  opacity={0}
  style={
    Object {
      "backgroundColor": "black",
      "bottom": 0,
      "display": "flex",
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 99,
    }
  }
/>
`;
